File: generated_code/python/solution_25_3.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # Time Complexity: O(N) where N is the number of nodes in the linked list. 
        # Each node is processed at most twice (once to count/check, once to reverse).
        # Space Complexity: O(1) auxiliary space as we modify pointers in place.
        
        # Base case: if list is empty or k is 1, no reversal needed.
        if not head or k == 1:
            return head
        
        # Use a dummy node to simplify handling the new head of the list.
        dummy = ListNode(0)
        dummy.next = head
        
        # group_prev will always point to the node just before the current k-group.
        group_prev = dummy
        
        while True:
            # Check if there are at least k nodes left to reverse.
            kth = group_prev
            for _ in range(k):
                kth = kth.next
                if not kth:
                    # Fewer than k nodes remaining, leave them as is and finish.
                    return dummy.next
            
            # Store the connection to the next group before reversing.
            group_next = kth.next
            
            # Reverse the current k-group.
            # prev starts at group_next so the tail of the reversed group connects to the next group.
            prev = group_next
            curr = group_prev.next
            
            for _ in range(k):
                nxt = curr.next
                curr.next = prev
                prev = curr
                curr = nxt
                
            # group_prev.next currently points to the old head of this group (which is now the tail).
            # We save it to become the next group_prev after updating connections.
            new_group_prev = group_prev.next
            
            # Connect the previous part of the list to the new head of the reversed group.
            group_prev.next = kth
            
            # Move group_prev to the tail of the current reversed group.
            group_prev = new_group_prev

File: generated_code/cpp/solution_25_3.cpp
#include <vector>
#include <algorithm>

// Definition for singly-linked list.
// Included for compilability as mandated by guidelines.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (!head || k == 1) return head;
        
        // Dummy node to simplify handling the new head of the list
        ListNode dummy(0);
        dummy.next = head;
        
        ListNode* prevGroupTail = &dummy;
        ListNode* curr = head;
        
        while (curr) {
            // Check if there are at least k nodes left to reverse
            ListNode* groupTail = prevGroupTail;
            for (int i = 0; i < k; ++i) {
                groupTail = groupTail->next;
                if (!groupTail) {
                    // Less than k nodes left, keep them as is and finish
                    return dummy.next;
                }
            }
            
            // Track the next group's head before we disconnect/reverse
            ListNode* nextGroupHead = groupTail->next;
            
            // Reverse the current group of k nodes
            ListNode* prev = nextGroupHead; // Connect the tail of reversed group to nextGroupHead
            ListNode* groupHead = curr;
            while (curr != nextGroupHead) {
                ListNode* nextNode = curr->next;
                curr->next = prev;
                prev = curr;
                curr = nextNode;
            }
            
            // Connect the previous group's tail to the new head of this reversed group
            prevGroupTail->next = prev;
            
            // The original head of this group is now the tail of this reversed group
            prevGroupTail = groupHead;
        }
        
        return dummy.next;
    }
};